SO-5834: generic dependency model

cmark opened the pull request.

cmark committed:
feat(core): introduce new dependency model

cmark committed:
feat(core): migrate resource creation to use the new dependency model

Old extensionOf and upgradeOf parameters are being converted to the new
model and persisted.
Support extensionOf/upgradeOf and the new dependency array in all
TerminologyResource subtypes.

cmark committed:
feat(core): allow updating new dependencies field through resource API

cmark committed:
feat(core): add missing setDependencies method to resource updates req

cmark committed:
Merge remote-tracking branch 'origin/8.x' into feature/SO-5834-dependencies-model

cmark committed:
fix(core): properly convert old dependency model values from new...

...dependencies array for backward compatibility

cmark committed:
feat(core): add hasQueryPart method to ResourceURIWithQuery

cmark committed:
feat(api): support `dependencies` when creating new codesystems

cmark committed:
feat(api): support `dependencies` when updating existing codesystems

apeteri commented on core/com.b2international.snowowl.core/src/com/b2international/snowowl/core/internal/DependencyDocument.java:
```diff
@@ -0,0 +1,85 @@
+/**
+ * @since 8.12.0
+ */
+public final class DependencyDocument implements Comparable<DependencyDocument> {
```
This class might need extra annotations or dummy setters for the derived fields if the "ignore unknown properties" option is not enabled.

cmark replied:
Tests are running just fine, so not sure about this.

cmark committed:
fix(test): revert upgrade API to use the old model until migration

cmark committed:
fix(core): properly update dependencies when syncing a codesystem...

...from a source URI

cmark committed:
feat(api): check dependency entry reference integrity before commit

cmark committed:
feat(api): check dependency entry duplication when creating a resource

cmark committed:
feat(api): check dependency entries when updating a resource

cmark committed:
feat(index): add support for query_string expression in index layer

cmark committed:
feat(core): rename Dependency field `resourceUri` to `uri`

cmark committed:
feat(core): support searching for dependency array entries via...

...query string syntax

Examples:
* ?dependency=codesystems/SNOMEDCT
* ?dependency=uri:codesystems/SNOMEDCT OR uri:codesystems/SNOMEDCT/*
* ?dependency=uri:codesystems/SNOMEDCT AND scope:extensionOf

cmark committed:
test(core): add test case to verify unversioned and versioned...

...dependency search

cmark committed:
feat(core): check if a resource being deleted is referenced by...

...other resources and prevent deletion unless `force` is set to true

cmark committed:
fix(core): scope compare in DependencyDocument

cmark committed:
test: remove duplicate codesystem api test

cmark committed:
feat(core): maintain dependencies on version documents for proper...

...versioned access

apeteri reviewed (approved):
LGTM!

cmark committed:
feat(core): minor Java API changes to ensure third party modules...

...work as intended

apeteri reviewed (approved):
👍

apeteri commented:
Does `snow-owl-ext` need a separate PR or is everything handled in classes of this project?

cmark commented:
> Does `snow-owl-ext` need a separate PR or is everything handled in classes of this project?

There is an open PR in that repo: https://github.com/b2ihealthcare/snow-owl-ext/pull/2681

cmark committed:
feat(core): make sure resource URIs with queries are recognized properly

AAAlinaaa commented on core/com.b2international.snowowl.core/src/com/b2international/snowowl/core/request/resource/BaseTerminologyResourceCreateRequest.java:
```diff
@@ -0,0 +1,388 @@
+			throw new BadRequestException("Branch path '%s' should point to an existing branch if given.", branchPath);
+		}
+		
+		// If the branch should be created, it branch should not exist, however 
```
typo: it branch

AAAlinaaa commented on core/com.b2international.snowowl.core/src/com/b2international/snowowl/core/request/resource/BaseTerminologyResourceUpdateRequest.java:
```diff
@@ -60,100 +85,118 @@ public final void setExtensionOf(ResourceURI extensionOf) {
+		}
+		
+		if (!oid.isBlank()) {
+			final boolean oidExist = ResourceRequests.prepareSearch()
```
Is an admin context not needed here?

cmark replied:
Yes and in other locations as well. (there is a Jira issue about it)

AAAlinaaa reviewed (approved):
LGTM

cmark committed:
fix(core): clarify comment in BaseTerminologyResourceCreateRequest

cmark committed:
feat(core): support updating dependencies through other parameters...

...like settings to make 8.12 backward compatible with pre-8.12 datasets
and clients.
